### PR TITLE
feature #856: added ability to update comments

### LINKED ISSUE #71: Add ability to update comments
One of the features of the admin panel will be editing of **TEXT** type answers. Your task is to create an endpoint to update a comment by its id. Also add a check so that it is exactly TEXT. Return an updated comment.

image

### LINKED ISSUE #856: Add ability to update comments
One of the features of the admin panel will be editing of **TEXT** type answers. Your task is to create an endpoint to update a comment by its id. Also add a check so that it is exactly TEXT. Return an updated comment.

image

### PR DESCRIPTION
**Pull request by illiakroshka**
- Created update in QuestionAnswerRepository
- Created updateComment in PollService
- Created PATCH teachers/comment/update endpoint
- Created UpdateCommentDTO
- Created InvalidTypeException
- Added permission
- Added Swagger documentation

closes fictadvisor/fictadvisor#71

### CODE REVIEW

File: src/v2/security/PERMISSION.ts [hunk 1]
```diff
@@ -86,6 +86,7 @@ export enum PERMISSION {
   USERS_$USERID_TEACHERS_$TEACHERID_DISCIPLINES_GET = 'users.$userId.teachers.$teacherId.disciplines.get',
 
   TEACHERS_CREATE = 'teachers.create',
+  TEACHERS_COMMENT_UPDATE = 'teachers.comment.update',
```
Comment (fokaaas) [inline src/v2/security/PERMISSION.ts]: disciplineTeachers.comment.update

File: src/v2/api/responses/UpdatedCommentResponse.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+export class UpdatedCommentResponse {
+  @ApiProperty({
+    description: 'Updated comment',
+  })
+    comment: string;
+}
```
Comment (fokaaas) [inline src/v2/api/responses/UpdatedCommentResponse.ts]: It must be ```class Comment``` as response

File: src/v2/api/controllers/DisciplineTeacherController.ts [hunk 3]
```diff
@@ -254,4 +260,55 @@ export class DisciplineTeacherController {
...
+  @ApiParam({
+    name: 'disciplineTeacherId',
+    required: true,
+    description: 'Discipline teacher id',
+  })
+  @ApiEndpoint({
+    summary: 'Update question answers with TEXT type (comments)',
+    permissions: PERMISSION.TEACHERS_COMMENT_UPDATE,
+  })
+  @Patch('/:disciplineTeacherId/update/comment')
```
Comment (fokaaas) [inline src/v2/api/controllers/DisciplineTeacherController.ts]: Don't specify in the path what the HTTP methods specify. Remove ```update```

File: src/v2/api/controllers/DisciplineTeacherController.ts [hunk 4]
```diff
@@ -254,4 +260,55 @@ export class DisciplineTeacherController {
...
+  async updateComment (
+    @Param('disciplineTeacherId', DisciplineTeacherByIdPipe) disciplineTeacherId: string,
+    @Body(UpdateCommentPipe) body: UpdateCommentDTO,
+  ) {
+    const updatedComment = await this.pollService.updateComment(disciplineTeacherId, body);
```
Comment (fokaaas) [inline src/v2/api/controllers/DisciplineTeacherController.ts]: Move logic to discipline teacher service.

**Review by fokaaas — CHANGES_REQUESTED** [review]

**Comment by illiakroshka** [comment]
It is not possible to have "comment" class in the responce because it doesn't return name of the discipline. That's why I created new custom UpdatedComment responce and also added question id there

File: src/v2/api/controllers/DisciplineTeacherController.ts [hunk 5]
```diff
@@ -254,4 +260,53 @@ export class DisciplineTeacherController {
...
+      You do not have permission to perform this action`,
+  })
+  @ApiParam({
+    name: 'disciplineTeacherId',
+    required: true,
+    description: 'Discipline teacher id',
+  })
+  @ApiEndpoint({
+    summary: 'Update question answers with TEXT type (comments)',
+    permissions: PERMISSION.DISCIPLINE_TEACHERS_COMMENT_UPDATE,
```
Comment (fokaaas) [inline src/v2/api/controllers/DisciplineTeacherController.ts]: just COMMENTS_UPDATE

File: src/v2/api/dtos/UpdateCommentDTO.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { IsNotEmpty, IsString } from 'class-validator';
+import { ApiProperty } from '@nestjs/swagger';
+
+export class UpdateCommentDTO {
+  @ApiProperty({
+    description: 'User\'s ID who provided the answer.',
+  })
+  @IsNotEmpty()
```
Comment (fokaaas) [inline src/v2/api/dtos/UpdateCommentDTO.ts]: add message

File: src/v2/api/dtos/UpdateCommentDTO.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    userId: string;
+
+  @ApiProperty({
+    description: 'Question id',
+  })
+  @IsNotEmpty()
```
Comment (fokaaas) [inline src/v2/api/dtos/UpdateCommentDTO.ts]: add message

File: src/v2/api/dtos/UpdateCommentDTO.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    questionId: string;
+
+  @ApiProperty({
+    description: 'New comment value',
+  })
+  @IsNotEmpty()
```
Comment (fokaaas) [inline src/v2/api/dtos/UpdateCommentDTO.ts]: add message

File: src/v2/api/pipes/UpdateCommentPipe.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+      throw new InvalidEntityIdException('User');
+    }
+
+    const question = await this.questionRepository.findById(body.questionId);
+    if (!question) {
+      throw new InvalidEntityIdException('Question');
+    }
+    if (question.type !== QuestionType.TEXT) {
+      throw new InvalidTypeException('Question');
+    }
```
Comment (fokaaas) [inline src/v2/api/pipes/UpdateCommentPipe.ts]: it's better not to create a pipe, transfer the business logic to the service

**Review by fokaaas — CHANGES_REQUESTED** [review]

**Review by fokaaas — APPROVED** [review]

File: src/v2/api/controllers/DisciplineTeacherController.ts [hunk 10]
```diff
@@ -254,4 +259,53 @@ export class DisciplineTeacherController {
...
+      User with such id is not found
+      Question with such id is not found
+      DisciplineTeacher with such id is not found
+      
+    InvalidTypeException
+      Question has wrong type 
+    
+    InvalidBodyException:
+      UserId should not be empty
+      DisciplineTeacherId should not be empty
```
Comment (SharpDevOps10) [inline src/v2/api/controllers/DisciplineTeacherController.ts]: Redundant exception  `DisciplineTeacherId should not be empty`

File: src/v2/api/dtos/UpdateCommentDTO.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  })
+  @IsNotEmpty(validationOptionsMsg('questionId should not be empty'))
+    questionId: string;
+
+  @ApiProperty({
+    description: 'New comment value',
+  })
+  @IsNotEmpty(validationOptionsMsg('comment should not be empty'))
+  @IsString(validationOptionsMsg('comment must be a string'))
+    comment: string;
```
Comment (SharpDevOps10) [inline src/v2/api/dtos/UpdateCommentDTO.ts]: validationOptionsMsg string parameter should start with a capital letter

**Review by SharpDevOps10 — CHANGES_REQUESTED** [review]
Also update `InvalidEntityIdException ('DisciplineTeacher')`

**Comment by fokaaas** [comment]
closed as outdated

**Pull request closed without merging**